src/ape/contracts/base.py

```diff
@@ -842,6 +843,35 @@ def at(self, address: AddressType, txn_hash: Optional[str] = None) -> ContractIn
             address, self.contract_type, txn_hash=txn_hash
         )
 
+    @cached_property
+    def source_path(self) -> Optional[Path]:
+        """
+        Returns the path to the local contract if determined that this container
+        belongs to the active project.
+        """
+        contract_name = self.contract_type.name
+        source_id = self.contract_type.source_id
+        if not (contract_name and source_id):
```

Comment:
should these have "warn" messages?

Reply:
No, it is valid a contract, just means it isn't in the project. There are various indicators of this. Mostly, we are just being very type-safe here.